### ISSUE TITLE
Replicate FID Scores

### ISSUE THREAD

**Issue by laurenzheidrich** [dexin-peng/score_sde_pytorch#7]
Hey!

First of all, thanks for this great work!

I am trying to replicate your FID scores below 3. I just generated 1K images using your sampling code and for the FID score it gives me:

2025-08-18 11:41:37,730 - runner - INFO - Calculating FID score using images from: ['assets/ve/cifar10_ncsnpp_cont/sample/1999/raw', 'data/CIFAR10.npz']
2025-08-18 11:41:44,799 - runner - INFO - FID Score: 32.1738112002796

This is more than 10x higher than what expected. So my question is, will that score dramatically reduce once I generate the 50K image that you did or should I expect to have a FID score in the range of 3 already with 1K generated samples.

Kinds regards!

**Comment by dexin-peng** [comment]
Hi Laurenz, Thanks for your attention!

The 1k images' FID is correct, you may check [this discussion](https://github.com/dexin-peng/score_sde_pytorch/issues/5#issuecomment-2969839949). When 50k images, the FID will be below 3.

**Comment by dexin-peng** [comment]
1. Prerequisite
You have generated 50k images and placed them in `./sample/ve/cifar10_nscnpp_cont/raw`

2. For rolling evaluate

```python3#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import shutil
import tempfile
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import torch
from pytorch_fid.fid_score import calculate_fid_given_paths


SAMPLE_DIR = "./sample/ve/cifar10_nscnpp_cont/raw"
DATA_DIR   = "./data"
REF_STAT   = os.path.join(DATA_DIR, "CIFAR10.npz")
REF_DIR    = os.path.join(DATA_DIR, "CIFAR10")


if os.path.isfile(REF_STAT):
    ref_path = REF_STAT
elif os.path.isdir(REF_DIR):
    ref_path = REF_DIR
else:
    raise FileNotFoundError("None of CIFAR10.npz nor CIFAR10 images dir is found")

step = 200
max_n = 50000

all_pngs = sorted(
    os.path.join(SAMPLE_DIR, f)
    for f in os.listdir(SAMPLE_DIR)
    if f.lower().endswith(".png")
)
n_total = min(len(all_pngs), max_n)

subset_dir = tempfile.mkdtemp(prefix="fid_subset_")
print(f"[INFO] subset_dir: {subset_dir}")

counts, scores = [], []
created = 0

try:
    for n in range(step, n_total + 1, step):
        for src in all_pngs[created:n]:
            name = f"{created:06d}.png"
            dst = os.path.join(subset_dir, name)
            try:
                os.symlink(src, dst)
            except (OSError, NotImplementedError):
                try:
                    os.link(src, dst)
                except OSError:
                    shutil.copy2(src, dst)
            created += 1

        fid = calculate_fid_given_paths(
            [subset_dir, ref_path],
            batch_size=50,
            device='cuda:0',
            dims=2048
        )
        counts.append(n)
        scores.append(float(fid))
        print(f"[FID] n={n:6d} -> {fid:.6f}")

finally:
    shutil.rmtree(subset_dir, ignore_errors=True)

s = pd.Series(scores, index=counts, name="FID")
s.index.name = "num_images"
s.to_csv("fid_progress.csv")

print("[DONE] saved fid_progress.csv")
```

3. For plotting

```python3
df = pd.read_csv('fid_progress.csv')
df = df.set_index('num_images')
ax = df.FID.plot()

# ax.xaxis.set_major_locator(mpl.ticker.MultipleLocator(2000))
# ax.xaxis.set_minor_locator(mpl.ticker.MultipleLocator(500))
ax.set_xlabel("Number of Generated Images", fontsize=12)
ax.set_ylabel("FID Score", fontsize=12)

special_xticks = [200, 1000, 3000, 50000]
for x in special_xticks:
    if x in df.index:
        y = df.loc[x, 'FID']
        ax.axvline(x=x, color='red', linestyle='--', alpha=0.6)
        ax.axhline(y=y, color='blue', linestyle='--', alpha=0.6)
        ax.scatter(x, y, color='black', zorder=5)
        ax.text(x, y, f"({x},{y:.2f})", fontsize=8, ha='right', va='bottom')

plt.grid(True, linestyle='--', linewidth=0.8, alpha=0.7, which='both')
plt.tight_layout()
plt.show()
```

4. Notes
To be consistent with what was claimed in the original paper, it should be 50k to 60k eval. Theoretically better setting may be 50k to 50k (I'm not quite sure, you may correct me on this), because the checkpoint is trained on CIFAR10 50k training images.

The paper claims 2.38 here, however, only 2.4646189212799 could be achieved with the same setting as the original repo
```python3
    #mean[:, 0, :, :] = 0.4914
    #mean[:, 1, :, :] = 0.4822
    #mean[:, 2, :, :] = 0.4465
```
If uncomment these lines in the `run/sde.py`, generate samples and eval
```python3
    mean[:, 0, :, :] = 0.4914
    mean[:, 1, :, :] = 0.4822
    mean[:, 2, :, :] = 0.4465
```

You will get 2.3975772857666 in the 50k to 60k eval, and the discrepancy is acceptable now
The blame for the discrepancy could be on the pretrained Inception nets, training epochs, RNG and the hardware arch
```
a) Trained on 60k images, instead of 50k
b) Inception-v3, instead of Inception-v1
c) Epoch: 2000(1999th)
d) Software: torch+cu version, cuBLAS version
e) Hardware: Computing Capability of cuda device
```

5. Tested settings (records from Apr 2025)

| Prior         | Trained on           | Generated Sample Num to CIFAR10 Num | Eval on      | FID Score        |
|---------------|----------------------|-------------------------------------|--------------|------------------|
| prior 0       | 60k CIFAR10 | 50k to 60k                          | Inception-v3 | 2.4646189212799  |
| correct prior | 60k CIFAR10 | 50k to 60k                          | Inception-v3 | 2.3975772857666  |
| correct prior | 60k CIFAR10 | 60k to 60k                          | Inception-v3 | 2.34573698043823 |


6. Controversial choices of this rewrite repo
a) For this generative task, it's safe to add the 10k CIFAR10 testset into training. And they were added. The provided `epoch_1999.pth` is trained including the 10k test samples
b) Eval is done with `Inception-v3`, instead of the v1, which is claimed in the original paper
c) The provided `./data/CIFAR10.npz` is the stats of 60k CIFAR10 images

7. Expected FID score
If you follow the default settings in this rewrite repo, you will get 2.465

Despite all these possible differences, you are still guaranteed to get an FID score 2.5±0.2)

Possible differences
a) 50k or 60k
b) Inception-v3 or Inception-v1
c) Prior distribution
d) Randomness
e) Hardware

**Comment by laurenzheidrich** [comment]
Thanks so much!! This is really helpful :)

Two further "changes" I have spotted compared to the original implementation are the following:

1) You set the warmup of the Adam optimizer = 2, in the original implementation I see that it is set to 5000. what is the reasoning for that?

2) You have introduced model.discretization steps and sampling.discretization_steps. Here I am wondering what the motivation for this is? Shouldn't that always be equal to the number of noise scales as long as there isn't any more elaborate sampling strategy?

Thanks in advance! :)
Laurenz

**Comment by laurenzheidrich** [comment]
Thanks!! Got another remark though:

I wanted to follow up on our previous discussion about scaling inputs to [-1,1]

While reviewing the code, I noticed that in line 49 of `train.py`:

`batch_data = self.data_loader.data_scaler(batch_data) `

the data is scaled to [-1, 1] if `self.config.data.centered` is set to `True`, which makes sense.

However, in the standard settings, `self.config.data.centered` is set to `False`, so this initial centering does not happen.

Then, in line 265 of `ncsnpp.py`, inside the `forward()` function, we have:

```
if not self.config.data.centered:
    x = 2 * x - 1.
```

So when centering is not applied during data loading, it gets applied inside the forward pass instead.

This raises two concerns:

	1.	If I disable centering in the config, I would expect no centering to take place anywhere, but here it still does in the forward pass.
	2.	More importantly, since the `forward()` method receives already noise-perturbed data (e.g.,  original image in [0,1], after noise perturbation in [-5, 6] then the data centering call in the forward function moves the image to [-11, 11]). 

While this centers the data, it also scales up the noise variance, which effectively changes the maximum noise level set in the config.

Would appreciate your thoughts on whether this behavior is intended or not!

Thanks!
Laurenz

**Comment by dexin-peng** [comment]
Great question!

This behavior is not intended. It's adopted from the [Line 259-261](https://github.com/yang-song/score_sde_pytorch/blob/cb1f359f4aadf0ff9a5e122fe8fffc9451fd6e44/models/ncsnpp.py#L259C1-L261C21) of the original repo's `./models/ncsnpp.py`

Yes, it will change the data distribution, even the input of the `forward()` function is bounded between -1 and 1 (and it's actually NOT bounded that way)

I got busy calculating the boundaries of the `2 * x - 1.`, which made my attention deviated when reviewing the Line 260 comments, and now it appears I was in smoke and not aware of the `data.centered` problem

Back to your concerns, in my opinion
1.	If centering is disabled in the config, no centering should take place anywhere, and it shouldn't be in the forward pass.
2.	The `x = 2 * x - 1.` in `forward()` method indeed perturbs the already noise-perturbed data.

To further confirm whether the "double-perturbation" is from the rewriting or from the original repo, check the original repo

The data is with the following pipeline
`batch -> train_step_fn in run_lib.py -> losses.get_step_fn -> get_sde_loss_fn -> loss_fn -> sde.marginal_prob`

Before the processed data is input into the ncsn++ model, focus on all possible linear transformations appeared in the code, found nothing works alike `x = (x + 1.) / 2`

Then it's confirmed that the double-perturbation is from the original repo

To confirm whether the double-perturbation is intended in the original paper, check it and find nothing looks like `center`, `-1`, `- 1`, indicating no theory supporting this behavior

Check other models in the original repo, found similar behavior in [Line 197-200](https://github.com/yang-song/score_sde_pytorch/blob/cb1f359f4aadf0ff9a5e122fe8fffc9451fd6e44/models/ncsnv2.py#L197C1-L200C12) of ncsnv2 definition

1. If there is no special concerns, in my opinion, the following codes (and the one in ncsnv2 definition) should be deleted, as it appears working in some aspects related to the `data.centered` specification according to the comments, and if centering is disabled in the config, no centering should take place anywhere

```python3
    if not self.config.data.centered:
      # If input data is in [0, 1]
      x = 2 * x - 1.
```

2. To make a fair comparison between ncsn series models and ddpm, which are all presented in the original paper, the "double-perturbation" should be removed. Otherwise it's not clear whether the extra gain of the ncsn series model is from its "ncsn series nature" or the special treatment of "double-perturbation"

3. The greatest contribution (in my opinion) of the original paper is unifying the two different threads of diffusion models, so it's acceptable to integrate the "double-perturbation" (making it more easy to recognize the noise and enhance the performance I guess?). And it's ok the explain the `x = 2 * x - 1.` as one of the built-in but unexplained mechanism of the "black box"  ncsn. (Though this explanation is not in my favor)
